### DISCUSSION TITLE
[Usability] Warning for wrong paramter type?

### DISCUSSION THREAD

**Discussion by LongnoseRob** [pymeasure/pymeasure#487] [Ideas]
I just ran in to an issue in one of my procedures where I assigned a float value to a `IntergerParameter` after changing the stepsize from 1dB steps to 0.1dB steps.
The procedure was running as normal, but looking at my result files made me relaize that something was not right.

Looking into the implementation I realized that the current implementation fo `IntegerParameter` does a cast to `int`
https://github.com/pymeasure/pymeasure/blob/f226ab4aaec8265ff442c5baadc27cfdee513ca4/pymeasure/experiment/parameters.py#L128-L138

maybe it would be nice to add a short check like this in line 129
`if type(value) is float:
   log.warning("float value found to be assigned to an IntegerParameter?!") `
to log a warning or trow an exception(`ValueError`)

Before I create a PR for this:

- would this be a benefit?
- give only a `log.warning` 
- raise a `ValueError`?

**Comment by bilderbuchi** [comment]
I'm not sure, given duck typing / the dynamic nature of Python, universally checking for this kind of error will probably be cumbersome in details. 
For example, if your `value` is `1.0`, there is no loss of precision, but your check will warn (I think an exception is too harsh as it would bring down the whole experiment).
Now, you could check for loss of precision using something like `int(value) == value`, but then what if someone passes in a string that converts cleanly to int, like `'1'`?
etc etc.
Also, be aware that some instruments are bound to return in float format even when the values are integral, which could also trigger warnings.

I think if a user assigns non-integral values to `IntegerParameter`s, we could make sure this is noticeable in the GUI? I.e, the widget only displays the whole number. I'm assuming this is already the case? Did you assign this value in code, not in the GUI, or how did that happen?

**Reply by LongnoseRob** [reply]
@bilderbuchi  Thank you for the comment/explanation.

I think my idea was good, but as you mention it would cost more efforts to implement then the actual uses.
For the GUI there might be an detection, I have not tried it yet.

_For the background how I came to this point:_
In my application I have two scripts one for the data collection and one for the analysis later.
Both are console scripts at the moment without GUI, the data collection is based on the `SImpleProcedure` example.

```
class SimpleProcedure(Procedure):
    Det_No = IntegerParameter('Det_No', default = 1)
    frequency = FloatParameter('Frequency', units= 'Hz',default=50E6)
    #level = IntegerParameter('Level', units= 'dBm', default=-10)
    level = FloatParameter('Level', units= 'dBm', default=-10)
    iterations = IntegerParameter('Loop Iterations', default=100)

    DATA_COLUMNS = ['Det_No','Frequency','Level','Iteration','Voltage']
    ...
```


Before I hade a simple list  assignment in the main part
```
level = [-30,-27,-24,-21,-18,-15,-12,-9,-6,-3,0,3,6,7,8,9,10]
``` 
But as I wanted to have more detailled data, I changed that to 
```
#0.1 dB steps
level = [i/10 for i in list(range(-330,101,1))]
```
With that I  came to the point where I assigned a float to the IntParameter, which I only noticed by checking some of my result-files and noticing that the level was not increasing in 0.1 steps.

Then I noticed my mistake of not updating the Parameter type at the beginning of the procedure.


The complete file is here:
https://gist.github.com/LongnoseRob/fe2e60085603ea94643a763a17eff49d#file-sft_det_measure-py
